File: python/main/nli/nli_generator.py
import hashlib
import tempfile
from datetime import datetime
from pathlib import Path
import shutil
from xml.dom.minidom import getDOMImplementation, Document, Element

import edrm
from edrm.DirectoryEntry import DirectoryEntry
from edrm.EDRMBuilder import EDRMBuilder
from edrm.EntryInterface import EntryInterface
from edrm.FileEntry import FileEntry
import edrm.EDRMUtilities as eutes


class NLIGenerator(object):
    def __init__(self):
        self.__edrm_builder = EDRMBuilder()
        self.__edrm_builder.as_nli = True

    def add_entry(self, entry: EntryInterface):
        if hasattr(entry, 'add_to_builder'):
            entry.add_to_builder(self.__edrm_builder)
        else:
            self.__edrm_builder.add_entry(entry)

    def generate_metadata_file(self, metadata_path: Path):
        metadata_file: Document = getDOMImplementation().createDocument(None, "image-metadata", None)
        property_list: Element = metadata_file.createElement('properties')

        case_element: Element = metadata_file.createElement('property')
        case_element.setAttribute("key", "case-number")
        case_element.setAttribute("value", "01")
        property_list.appendChild(case_element)

        datetime_element: Element = metadata_file.createElement('property')
        datetime_element.setAttribute("key", "creation-datetime")
        datetime_element.setAttribute("value", datetime.now().strftime('%Y/%m/%d %H:%M:%S.%f')[:-3] + " UTC")
        property_list.appendChild(datetime_element)

        sw_element: Element = metadata_file.createElement('property')
        sw_element.setAttribute("key", "creation-software-name")
        sw_element.setAttribute("value", "Nuix Memory Analysis Tool")
        property_list.appendChild(sw_element)

        swv_element: Element = metadata_file.createElement('property')
        swv_element.setAttribute("key", "creation-software-version")
        swv_element.setAttribute("value", "0.0.1")
        property_list.appendChild(swv_element)

        evidence_element: Element = metadata_file.createElement('property')
        evidence_element.setAttribute("key", "evidence-number")
        evidence_element.setAttribute("value", "01")
        property_list.appendChild(evidence_element)

        examiner_element: Element = metadata_file.createElement('property')
        examiner_element.setAttribute("key", "examiner-name")
        examiner_element.setAttribute("value", "Unknown")
        property_list.appendChild(examiner_element)

        metadata_file.documentElement.appendChild(property_list)

        metadata_file_path: Path = metadata_path / 'image_metadata.xml'
        with metadata_file_path.open(mode='w', encoding=edrm.configs['encoding']) as metadata_xml:
            metadata_file.writexml(metadata_xml, encoding=edrm.configs['encoding'], addindent='    ', newl='\n')

    def save(self, file_path: Path):
        with tempfile.TemporaryDirectory(delete=False) as temp_loc:
            temp_path = Path(temp_loc)
            build_path = temp_path / 'NLI_Gen'
            metadata_path = build_path / '.metadata'
            metadata_path.mkdir(parents=True, exist_ok=True)

            # Make the EDRM XML 1.2 file
            self.__edrm_builder.output_path = metadata_path / 'image_contents.xml'
            self.__edrm_builder.save()

            # Copy files and folders to their temp location
            entry_map = self.__edrm_builder.entry_map
            for entry in self.__edrm_builder.entry_map.values():
                if isinstance(entry, FileEntry):
                    relative_path = eutes.generate_relative_path(entry, entry_map)
                    destination_path = build_path / relative_path

                    if destination_path.parent is not None:
                        destination_path.parent.mkdir(parents=True, exist_ok=True)

                    if isinstance(entry, DirectoryEntry):
                        destination_path.mkdir(exist_ok=True)
                    else:
                        shutil.copy2(entry.file_path, destination_path)

            # make the .metadata/image_metadata.xml file
            self.generate_metadata_file(metadata_path)

            # make the .metadata/image_contents.sha1_hash file
            metadata_hash = edrm.EDRMUtilities.hash_file(self.__edrm_builder.output_path,
                                                         hashlib.sha1(),
                                                         as_string=False)
            metadata_hash_path = metadata_path / 'image_contents.sha1_hash'
            with metadata_hash_path.open(mode='wb') as metadata_hash_file:
                metadata_hash_file.write(metadata_hash)

            # Zip the contents
            temp_nli_path = temp_path / f'{file_path.stem}'
            shutil.make_archive(str(temp_nli_path), 'zip', root_dir=build_path)

            # Copy to file_path
            shutil.copy(f'{temp_nli_path}.zip', str(file_path))

File: java/src/main/java/com/nuix/nli/nli/NLIGenerator.java
package com.nuix.nli.nli;

import com.nuix.nli.edrm.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.*;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Map;
import java.util.Properties;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class NLIGenerator {
    private static final boolean IS_WINDOWS = System.getProperty("os.name").toLowerCase().contains("win");

    private static Map<String, String> loadNLIConfig() {
        Properties props = new Properties();
        try (InputStream input = NLIGenerator.class.getResourceAsStream("/nli.config")) {
            if (input == null) {
                throw new RuntimeException("Could not find nli.config in resources");
            }
            props.load(input);
            return props.entrySet().stream()
                    .collect(Collectors.toMap(
                            e -> String.valueOf(e.getKey()),
                            e -> String.valueOf(e.getValue())
                    ));
        } catch (IOException e) {
            throw new RuntimeException("Failed to load nli.config", e);
        }
    }
    public static final Map<String,String> NLI_CONFIG = loadNLIConfig();

    private final EDRMBuilder builder = new EDRMBuilder();

    public NLIGenerator() {
        builder.setAsNli(true);
    }

    public String addEntry(EntryInterface entry) {
        if (entry instanceof CompoundEnty ce) {
            return ce.addToBuilder(builder);
        } else {
            return builder.addEntry(entry);
        }
    }
    public EntryInterface getEntry(String id) {
        return builder.getEntry(id);
    }

    public String addFile(String filePath, String mimeType, String parentId) { return builder.addFile(filePath, mimeType, parentId); }
    public String addDirectory(String directoryPath, String parentId) { return builder.addDirectory(directoryPath, parentId); }
    public String addMapping(Map<String, Object> mapping, String mimeType, String parentId) { return builder.addMapping(mapping, mimeType, parentId); }

    private void generateMetadataFile(Path metadataPath) {
        try {
            Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element root = doc.createElement("image-metadata");
            doc.appendChild(root);

            Element properties = doc.createElement("properties");
            root.appendChild(properties);

            Element caseElement = doc.createElement("property");
            caseElement.setAttribute("key", "case-number");
            caseElement.setAttribute("value", "01");
            properties.appendChild(caseElement);

            Element dateTimeElement = doc.createElement("property");
            dateTimeElement.setAttribute("key", "creation-datetime");
            dateTimeElement.setAttribute("value", OffsetDateTime.now().format(DateTimeFormatter.ofPattern(EDRMUtilities.EDRM_CONFIG.getOrDefault("date_time_format", "yyyy-MM-dd'T'HH:mm:ss.SSS"))) + " UTC");
            properties.appendChild(dateTimeElement);

            Element softwareElement = doc.createElement("property");
            softwareElement.setAttribute("key", "creation-software-name");
            softwareElement.setAttribute("value", "Nuix NLI Builder");

            Element evidenceElement = doc.createElement("property");
            evidenceElement.setAttribute("key", "evidence-number");
            evidenceElement.setAttribute("value", "01");
            properties.appendChild(evidenceElement);

            Element examinerElement = doc.createElement("property");
            examinerElement.setAttribute("key", "examiner-name");
            examinerElement.setAttribute("value", "Unknown");
            properties.appendChild(examinerElement);

            Path outputPath = metadataPath.resolve("image_metadata.xml");
            Files.createDirectories(outputPath.getParent());
            Transformer t = TransformerFactory.newInstance().newTransformer();
            t.setOutputProperty(OutputKeys.INDENT, "yes");
            t.setOutputProperty(OutputKeys.METHOD, "xml");
            t.setOutputProperty(OutputKeys.ENCODING, EDRMUtilities.EDRM_CONFIG.getOrDefault("encoding", "UTF-8"));
            t.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
            t.transform(new DOMSource(doc), new StreamResult(outputPath.toFile()));
        } catch (ParserConfigurationException | IOException | javax.xml.transform.TransformerException e) {
            throw new RuntimeException(e);
        }
    }

    public void save(Path nliPath) {
        boolean doDelete = !Boolean.parseBoolean(NLI_CONFIG.getOrDefault("debug", "false"));

        try {
            Path temp = Files.createTempDirectory("nli_gen");
            try {
                Path buildPath = temp.resolve("NLI_Gen");
                Path metadataPath = buildPath.resolve("._metadata");
                Files.createDirectories(metadataPath);

                // Save EDRM XML
                Path edrmXml = metadataPath.resolve("image_contents.xml");
                builder.setOutputPath(edrmXml);
                builder.save();

                Map<String, EntryInterface> entryMap = builder.getEntryMap();
                for (EntryInterface entry : entryMap.values()) {
                    if (entry instanceof FileEntry fe) {
                        String relativePath;
                        if (null == fe.getParentId()) {
                            relativePath = fe.getFilePath().getFileName().toString();
                        } else if (entryMap.get(fe.getParentId()) instanceof DirectoryEntry de) {
                            relativePath = EDRMUtilities.generateRelativePath(de, entryMap);
                            relativePath = relativePath + "/" + fe.getFilePath().getFileName().toString();
                        } else {
                            relativePath = Path.of("natives", fe.getFilePath().getFileName().toString()).toString();
                        }

                        Path destinationPath = buildPath.resolve(relativePath);
                        if (null != destinationPath.getParent()) {
                            Files.createDirectories(destinationPath.getParent());
                        }

                        if (fe instanceof DirectoryEntry de) {
                            Files.createDirectories(destinationPath);
                        } else {
                            Files.copy(fe.getFilePath(), destinationPath, StandardCopyOption.REPLACE_EXISTING);
                        }
                    } else if (entry instanceof MappingEntry me) {
                        Path mappingPath;
                        if (null != me.getText()) {
                            mappingPath = buildPath.resolve("natives").resolve(me.getField(me.getIdentifierField()).getValue().toString());
                            mappingPath = mappingPath.normalize().toAbsolutePath();
                            Files.createDirectories(mappingPath.getParent());
                            if (IS_WINDOWS) {
                                mappingPath = Path.of("\\\\?\\" + mappingPath.toString());
                            }
                            Files.writeString(mappingPath, me.getText(),
                                    java.nio.charset.Charset.forName(EDRMUtilities.EDRM_CONFIG.getOrDefault("encoding", "UTF-8")));
                        }
                    }
                }

                generateMetadataFile(metadataPath);

                byte[] metadataHash = EDRMUtilities.hashFile(builder.getOutputPath(), "SHA-1").hashBytes();
                Path metadataHashPath = metadataPath.resolve("image_contents.sha1_hash");
                Files.write(metadataHashPath, metadataHash);

                // Zip the contents
                Path zipPath = temp.resolve(NLIUtilities.getFileNameWithoutExtension(nliPath) + ".zip");
                try (ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(zipPath))) {
                    try(Stream<Path> walked = Files.walk(buildPath)) {
                        walked.filter(Files::isRegularFile).forEach(p -> {
                            try {
                                String entryName = buildPath.relativize(p).toString().replace('\\', '/');
                                zos.putNextEntry(new ZipEntry(entryName));
                                Files.copy(p, zos);
                                zos.closeEntry();
                            } catch (IOException e) { throw new RuntimeException(e); }
                        });
                    }
                }

                Files.copy(zipPath, nliPath, StandardCopyOption.REPLACE_EXISTING);
            } finally {
                if (doDelete) {
                    try {
                        NLIUtilities.deleteRecursively(temp);
                    } catch (IOException e) {
                        // Ignore delete exceptions
                    }
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}